osu.Game.Rulesets.Osu/Mods/OsuModDepth.cs

```diff
@@ -0,0 +1,138 @@
+            MinValue = 50,
+            MaxValue = 200
+        };
+
+        [SettingSource("Show Approach Circles", "Whether approach circles should be visible.", 1)]
+        public BindableBool ShowApproachCircles { get; } = new BindableBool(true);
+
+        protected override void ApplyIncreasedVisibilityState(DrawableHitObject hitObject, ArmedState state) => applyTransform(hitObject, state);
+
+        protected override void ApplyNormalVisibilityState(DrawableHitObject hitObject, ArmedState state) => applyTransform(hitObject, state);
+
+        public void ApplyToDrawableRuleset(DrawableRuleset<OsuHitObject> drawableRuleset)
+        {
+            // Hide judgment displays and follow points as they won't make any sense.
+            // Judgements can potentially be turned on in a future where they display at a position relative to their drawable counterpart.
+            drawableRuleset.Playfield.DisplayJudgements.Value = false;
+            (drawableRuleset.Playfield as OsuPlayfield)?.FollowPoints.Hide();
+        }
+
+        private void applyTransform(DrawableHitObject drawable, ArmedState state)
+        {
+            switch (drawable)
+            {
+                case DrawableHitCircle circle:
+                    if (!ShowApproachCircles.Value)
+                    {
+                        var hitObject = (OsuHitObject)drawable.HitObject;
+                        double appearTime = hitObject.StartTime - hitObject.TimePreempt;
+
+                        using (circle.BeginAbsoluteSequence(appearTime))
+                            circle.ApproachCircle.Hide();
+                    }
+
+                    break;
+            }
+        }
+
+        public void Update(Playfield playfield)
+        {
+            double time = playfield.Time.Current;
+
+            foreach (var drawable in playfield.HitObjectContainer.AliveObjects)
+            {
+                switch (drawable)
+                {
+                    case DrawableHitCircle circle:
+                        processObject(time, circle, 0);
+                        break;
+
+                    case DrawableSlider slider:
+                        processObject(time, slider, slider.HitObject.Duration);
+                        break;
+                }
+            }
+        }
+
+        private void processObject(double time, DrawableOsuHitObject drawable, double duration)
+        {
+            var hitObject = drawable.HitObject;
+
+            double baseSpeed = MaxDepth.Value / hitObject.TimePreempt;
+            double offsetAfterStartTime = duration + hitObject.MaximumJudgementOffset + 500;
+            double slowSpeed = Math.Min(-minDepth / offsetAfterStartTime, baseSpeed);
+
+            double decelerationTime = hitObject.TimePreempt * 0.2;
+            float decelerationDistance = (float)(decelerationTime * (baseSpeed + slowSpeed) * 0.5);
+
+            float z;
+
+            if (time < hitObject.StartTime - decelerationTime)
```

Comment:
I dunno, this feels worse to me than not having it at all.

```diff
diff --git a/osu.Game.Rulesets.Osu/Mods/OsuModDepth.cs b/osu.Game.Rulesets.Osu/Mods/OsuModDepth.cs
index 18d4fef5e8..16517a2f36 100644
--- a/osu.Game.Rulesets.Osu/Mods/OsuModDepth.cs
+++ b/osu.Game.Rulesets.Osu/Mods/OsuModDepth.cs
@@ -103,13 +103,7 @@ private void processObject(double time, DrawableOsuHitObject drawable, double du
 
             float z;
 
-            if (time < hitObject.StartTime - decelerationTime)
-            {
-                double appearTime = hitObject.StartTime - hitObject.TimePreempt;
-                float fullDistance = decelerationDistance + (float)(baseSpeed * (hitObject.TimePreempt - decelerationTime));
-                z = fullDistance - (float)((Math.Max(time, appearTime) - appearTime) * baseSpeed);
-            }
-            else if (time < hitObject.StartTime)
+            if (time < hitObject.StartTime)
             {
                 double timeOffset = time - (hitObject.StartTime - decelerationTime);
                 double deceleration = (slowSpeed - baseSpeed) / decelerationTime;

```

Reply:
Personally I prefer my version more, but may be I'm just biased